pkg/handlers/generic/lifecycle/serviceloadbalancer/metallb/handler.go

```diff
@@ -152,5 +158,77 @@ func (n *MetalLB) Apply(
 		return fmt.Errorf("failed to apply MetalLB installation HelmChartProxy: %w", err)
 	}
 
+	if err := wait.ForObject(
+		ctx,
+		wait.ForObjectInput[*caaphv1.HelmChartProxy]{
+			Reader: n.client,
+			Target: hcp.DeepCopy(),
+			Check: func(_ context.Context, obj *caaphv1.HelmChartProxy) (bool, error) {
+				for _, c := range obj.GetConditions() {
+					if c.Type == caaphv1.HelmReleaseProxiesReadyCondition && c.Status == corev1.ConditionTrue {
+						return true, nil
+					}
+				}
+				return false, nil
+			},
+			Interval: 5 * time.Second,
+			Timeout:  10 * time.Second,
```

Comment:
In the PR description you mentioned 20 seconds.

Reply:
Good observation.

We want the handler to return within 30 seconds. We should let the context determine that; the handler caller should cancel the context in 30 seconds. Each wait can be increased to 30 seconds; they will be cancelled as soon as the handler context is cancelled. I'll make this change.

Reply:
The context originates here: https://github.com/kubernetes-sigs/cluster-api/blob/c05b21b5f8710e4171975b1fe57fca68933fdd6f/exp/runtime/server/server.go#L315

It comes from the request, and the godoc says:

> // For incoming server requests, the context is canceled when the
> // client's connection closes, the request is canceled (with HTTP/2),
> // or when the ServeHTTP method returns.

CAPI says:

> Runtime Extension developers can decide how long the Cluster API Runtime should wait for a Runtime Extension to respond before treating the call as a failure (max is 30s) by returning the timeout during discovery. 
> -- https://cluster-api.sigs.k8s.io/tasks/experimental-features/runtime-sdk/implement-extensions.html#timeouts

I guess the client cancels the connection in 30s, triggering the context cancellation.

Reply:
Thanks for explaining!